### ISSUE TITLE
A11YMS: Accessibility Manager Service

### ISSUE THREAD

**Issue by clixyz** [clixyz/blog-notes#111]
## Overview

- src: /frameworks/base/services/accessibility/
- A11YMS is created and started as an other service by system server
  - A11YMS runs in the main thread of system server (message handling)
  - A11YMS registers broadcast receivers for broadcast and updates user states accordingly
    - ACTION_USER_SWITCHED
    - ACTION_USER_UNLOCKED
    - ... 
  - A11YMS observes content changes using a content observer and updates user states accordingly
    - TOUCH_EXPLORATION_ENABLED
    - ACCESSIBILITY_DISPLAY_MAGNIFICATION_ENABLED
    - ...
- A11YMS manages all a11y services either a real a11y service or a fake service (UiAutomation) of each device user, specifically, it saves them in the UserState of each userId (`A11YMS.mUserStates`)
- {add,remove}AccessibilityInteractionConnection(IWindow windowToken, IAccessibilityInteractionConnection connection, int userId)
  - add/remove an IAccessibilityInteractionConnection from a window, and saves the connection to state of userId
- [un]registerUiTestAutomationService(IBinder owner, IAccessibilityServiceClient serviceClient, AccessibilityServiceInfo accessibilityServiceInfo, int flags)
  - register/unregister an client from UiAutomation, create and save a fake Service (namely foo.bar/.FakeService) for it, invoke back to serviceClient that the service connection is created
  - each owner should register one and only UiAutomation
- sendAccessibilityEvent
- getInstalledAccessibilityServiceList
- getEnabledAccessibilityServiceList

> #### PackageMonitor
> - Helper class for monitoring the state of packages: adding, removing, updating, and disappearing and reappearing on the SD card 
> - Actually is a wrapper class of registering broadcast receivers for broadcasts that are relevant with packages
>
> #### ContentObserver
> - Receives call backs for changes to content

## Interaction among app, A11YMS, and UiAutomation

``` kotlin
val ua = UiAutomation(looper, UiAutomationConnection()).apply {
  connect();
}
val root = ua.rootInActiveWindow
```

#### TLDR; 

UiAutomation get the app view root by querying A11YMS, and A11YMS will query the app to help fetch the app view hierarchy

The three runs inside three processes, i.e., the app process, the system server process, and the process who uses UiAutomation. The commnunication among them is via, apparently, binder. 
- Each app is a binder server that serves the a11y view hierarchy
- A11YMS is a binder server that manages all above binder servers
- UiAutomation is a A11YMS client that asks A11YMS for the view hierarchy

#### App

Each app exposes itself as an `IAccessibilityInteractionConnection` binder server. When decor view of each app's activity is set to its ViewRootImpl via `ViewRootImpl#setView(view)`, ViewRootImpl starts the binder server and register itself to A11YMS. 

``` java
public final class ViewRootImpl implements ViewParent,
  View.AttachInfo.Callbacks, ThreadedRenderer.DrawCallbacks { 

  public void setView(View view, WindowManager.LayoutParams attrs, View panelParentView) {
    ...

    // ensure the connection to A11YMS, see following
    mAccessibilityInteractionConnectionManager.ensureConnection();

    ...
  }
}

final class AccessibilityInteractionConnectionManager 
  implements AccessibilityStateChangeListener {
  
  public void ensureConnection() {
    ...

    // add self as an AccessibilityInteractionConnection to A11YMS
    mAttachInfo.mAccessibilityWindowId = 
      mAccessibilityManager.addAccessibilityInteractionConnection(
        mWindow,
        new AccessibilityInteractionConnection(ViewRootImpl.this));

    ...
  }	
}
```

#### A11YMS (system server process)

When the binder communication `addAccessibilityInteractionConnection()` from the app process comes, A11YMS checks the userId and add the connection to either global or the UserState of the userId

```
public int addAccessibilityInteractionConnection(IWindow windowToken,
  IAccessibilityInteractionConnection connection, int userId) throws RemoteException {

  if (mSecurityPolicy.isCallerInteractingAcrossUsers(userId)) {
    // add to global
    AccessibilityConnectionWrapper wrapper = new AccessibilityConnectionWrapper(
      windowId, connection, UserHandle.USER_ALL);
    wrapper.linkToDeath();
    mGlobalInteractionConnections.put(windowId, wrapper);
    mGlobalWindowTokens.put(windowId, windowToken.asBinder());
  } else {
    // add to userId
    AccessibilityConnectionWrapper wrapper = new AccessibilityConnectionWrapper(
      windowId, connection, resolvedUserId);
    wrapper.linkToDeath();
    UserState userState = getUserStateLocked(resolvedUserId);
    userState.mInteractionConnections.put(windowId, wrapper);
    userState.mWindowTokens.put(windowId, windowToken.asBinder());
  }
}
```

#### UiAutomation

When an UiAutomation is created, and connected, it starts itself as an `IAccessibilityServiceClientImpl extends IAccessibilityServiceClientWrapper extends IAccessibilityServiceClient.Stub` binder server, registers itself to A11YMS, and asynchrously waits for the connection get back (IAccessibilityServiceClient#init() will be called back)

And `val ua = UiAutomation(looper, UiAutomationConnection()).apply { connect() }` is equivalent to

``` java
IBinder token = new Binder();
// create the client
IAccessibilityServiceClient client = new IAccessibilityServiceClientImpl(looper);
// set the info that the client want
AccessibilityServiceInfo info = new AccessibilityServiceInfo();
// ... set info flags
IAccessibilityManager a11yms = IAccessibilityManager.Stub.asInterface(ServiceManager.getService(Context.ACCESSIBILITY_SERVICE));
// register the client
a11yms.registerUiTestAutomationService(token, client, info, flags)
```

#### A11YMS

When the binder communication `registerUiTestAutomationService()` from UiAutomation comes, A11YMS set UserState of the userId by
- giving the service a fake component name `foo.bar/FakeService` (i.e., `A11YMS.sFakeAccessibilityServiceComponentName`)
- saving the service client (each userId can have one and only one client)
- adding the service as an installed and enabled service
- notifying user state changed

``` java
public void registerUiTestAutomationService(IBinder owner,
  IAccessibilityServiceClient serviceClient,
  AccessibilityServiceInfo accessibilityServiceInfo,
  int flags) {
    
    // ensure only one client
    if (userState.mUiAutomationService != null) {
        throw new IllegalStateException("UiAutomationService " + serviceClient
                + "already registered!");
    }

    // give it a fake name for creating a fake service
    accessibilityServiceInfo.setComponentName(sFakeAccessibilityServiceComponentName);

    // save the client
    userState.mUiAutomationServiceOwner = owner;
    userState.mUiAutomationServiceClient = serviceClient;
    userState.mUiAutomationFlags = flags;

    // adding and enabling
    userState.mInstalledServices.add(accessibilityServiceInfo);
    userState.mEnabledServices.add(sFakeAccessibilityServiceComponentName);

    // notify user state changed
    onUserStateChangedLocked(userState);
}
```

Whenever the user state is changed, the service infos (installed, enabled, bound) are updated. Specifically, for each installed and enabled service
- ensure an `A11YMS#Service extends IAccessibilityServiceConnection.Stub` binders server (service) is created for it
- ensure the binder server is bound
  - save the service to the UserState via its fake name
  - callback to the `IAccessibilitySerivceClient#init(service)` that the connection is created

``` java
private void onUserStateChangedLocked(UserState userState) {
  updateServicesLocked(userState);
}

private void updateServicesLocked(UserState userState) {
  // ...

  for (int i = 0, count = userState.mInstalledServices.size(); i < count; i++) {
    AccessibilityServiceInfo installedService = userState.mInstalledServices.get(i);
    ComponentName componentName = ComponentName.unflattenFromString(installedService.getId());
    Service service = componentNameToServiceMap.get(componentName);

    // ...

    if (userState.mEnabledServices.contains(componentName)) {
      if (service == null) {
        // create a service
        service = new Service(userState.mUserId, componentName, installedService);
      } else if (userState.mBoundServices.contains(service)) {
        continue;
      }
      // add the service bind the service
      service.bindLocked();
    } else {
      if (service != null) {
        service.unbindLocked();
      }
    }
  }
}

class Service extends IAccessibilityServiceConnection.Stub {

  public boolean bindLocked() {
    mMainHandler.post(new Runnable() {
        @Override
        public void run() {
            // Simulate asynchronous connection since in onServiceConnected
            // we may modify the state data in case of an error but bind is
            // called while iterating over the data and bad things can happen.
            onServiceConnected(mComponentName,
                    userState.mUiAutomationServiceClient.asBinder());
        }
    });
  }

  public void onServiceConnected(ComponentName componentName, IBinder service) {
    // add service to UserState
    addServiceLocked(this, userState);

    // get the service client, service here is `userState.mUiAutomationServiceClient.asBinder()`
    mServiceInterface = IAccessibilityServiceClient.Stub.asInterface(service);

    // callback init() to notify the connection is created
    mServiceInterface.init(this, mId, mOverlayWindowToken);
  }
}
```

#### UiAutomation

When `init(connection)` is called back, the client send a `DO_INIT` message to its looper. When execute this message, it saves the connection and its id to its global singleton `AccessibilityInteractionClient`

``` java
public static class IAccessibilityServiceClientWrapper extends IAccessibilityServiceClient.Stub
  implements HandlerCaller.Callback {

  public void init(IAccessibilityServiceConnection connection, int connectionId,
    IBinder windowToken) {
    // send a DO_INIT message, mCaller = new HandlerCaller(context, looper, this, true /*asyncHandler*/);
    Message message = mCaller.obtainMessageIOO(DO_INIT, connectionId,
      connection, windowToken);
    mCaller.sendMessage(message);
  }

  @Override
  public void executeMessage(Message message) {
    case DO_INIT: {
      IAccessibilityServiceConnection connection = (IAccessibilityServiceConnection) args.arg1;
      // save to AccessibilityInteractionClient
      AccessibilityInteractionClient.getInstance().addConnection(mConnectionId, connection);
    }
  }
}
```

#### Relationships of app, A11YMS, and UiAutomation

```

                                      .-------------app Process------------.
                                      | server: A11YInteractionController<-+---.
                                      `------------------------------------'   |
                                                                               |
                                                                               ^ ^^ A11YMS as a client of app
                                                                               |    to manage and fetch its a11y 
                                                                               |    view hierarchy
                                                                               |
                                    .------------A11YMS Process-------------.  ^
                                 .--+>as server: IA11YServiceConnection     |  |
                                 |  |---------------------------------------|  |
                                 |  | as client: A11YInteractionController--+--'
                                 |  |---------------------------------------|
                                 ^  | as client: IA11YServiceClient---->----+--.
                                 |  `---------------------------------------'  |
                                 |                                             |
                                 |                                             |
  UiAutomation as a client of ^^ ^                                             v vv A11YMS as a client of UiAutomation 
 A11YMS to delegate its query    |                                             |    to manage and callback
                    to A11YMS    |                                             |
                                 |                                             v
                                 ^                                             |
                                 |    .--------UiAutomation Process-------.    |
                                 |    | as server: IA11YServiceClient<----+-<--'
                                 |    |-----------------------------------|
                                 `--<-+-as client: IA11YServiceConnection |
                                      `-----------------------------------'
 
```

#### UiAutomation.rootInActiveWindow

To fetch to root view, UiAutomation delegate its qeury to A11YMS by its connection, then A11YMS will query app the root view via his controller. 

The method invocation chain is
```
   (UiAutomation Process) UiAutomation.rootInActiveWindow
-> (UiAutomation Process) AccessibilityInteractionClient#getRootInActiveWindow(connectionId) (the singleton)
-> (UiAutomation Process) AccessibilityInteractionClient#findAccessibilityNodeInfoByAccessibilityId(connectionId, ACTIVE_WINDOW, ROOT_NODE)
-> (UiAutomation Process) IAccessibilityServiceConnection#findAccessibilityNodeInfoByAccessibilityId(connectionId, ACTIVE_WINDOW, ROOT_NODE)
-> (A11YMS Process) Service#findAccessibilityNodeInfoByAccessibilityId(connectionId, ACTIVE_WINDOW, ROOT_NODE)
-> (A11YMS Process) IAccessibilityInteractionConnection#findAccessibilityNodeInfoByAccessibilityId(connectionId, ACTIVE_WINDOW, ROOT_NODE)
-> (app Process) AccessibilityInteractionConnection#findAccessibilityNodeInfoByAccessibilityId(connectionId, ACTIVE_WINDOW, ROOT_NODE)
-> (app Process) AccessibilityInteractionController#findAccessibilityNodeInfoByAccessibilityIdClientThread(connectionId, ACTIVE_WINDOW, ROOT_NODE, currentThread)
-> (app Process) PrivateHandler#handleMessage() of message type PrivateHandler.MSG_FIND_ACCESSIBILITY_NODE_INFOS_BY_VIEW_ID
-> (app Process) AccessibilityInteractionController#findAccessibilityNodeInfoByAccessibilityIdUiThread(connectionId, ACTIVE_WINDOW, ROOT_NODE)
-> (app Process) AccessibilityNodePrefetcher#prefetchAccessibilityNodeInfos(rootView)
-> (app Process) View#createAccessibilityNodeInfo()
-> (app Process) View#createAccessibilityNodeInfoInternal()
-> (app Process) View#onInitializeAccessibilityNodeInfo()
```
